### PR TITLE
[#6645] Platform: AZ rendering issue when API throws bad request in case of same instance type. 

### PR DESCRIPTION
**Pull request by gaurav061**
Description: 
1. API is returning bad request when the instance type is changed and changed back to its original value. This was causing the Az rendering issue. The issue is addressed by calling fetch universe by UUID in case of an error in config API.
2. Blocked check universe API call when any field is changed other than universe name.

Test Plan: 
1. Changed instance type to something other than default and changed it back to its original value, AZ was rendered as expected.
2. Created a Universe of various types, TLS Enabled, TLS Disabled, c5.xlarge universe, on-prem Universe.

Current existing API find by name is given 4xx in case universe name is not exist and 200 in case universe exist but front end behavior is exactly opposite of it. The front end wants API which required to give 200 and allow the user to create a universe in case the universe name does not exist in the system. If the universe exists then We required to show an error message. incase we use to find universe API in the front end we required broke best practice rule. To fix that added one more API called check_unique which will give 200 in case of the universe not exist and 4xx in the case already there.

[info] Test run finished: 0 failed, 0 ignored, 107 total, 117.85s
[info] Passed: Total 107, Failed 0, Errors 0, Passed 107
[success] Total time: 166 s (02:46), completed 24 May, 2021 3:03:25 PM

### CODE REVIEW

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js [hunk 1]
```diff
@@ -61,9 +61,14 @@ import { toast } from 'react-toastify';
 
 const mapDispatchToProps = (dispatch) => {
   return {
-    submitConfigureUniverse: (values) => {
+    submitConfigureUniverse: (values, uuid = null) => {
```
Comment (nishantSharma459) [inline managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js:64]: `uuid = null` you can do it in the component only.

File: managed/ui/src/components/universes/UniverseForm/ClusterFields.js [hunk 2]
```diff
@@ -928,7 +928,7 @@ export default class ClusterFields extends Component {
       formValues,
       clusterType
     } = this.props;
-
+
```
Comment (nishantSharma459) [inline managed/ui/src/components/universes/UniverseForm/ClusterFields.js:931]: Indent

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js [hunk 3]
```diff
@@ -521,7 +526,8 @@ const universeForm = reduxForm({
   form: 'UniverseForm',
   validate,
   asyncValidate,
-  fields: formFieldNames
+  fields: formFieldNames,
+  asyncChangeFields: ['primary.universeName', 'async.universeName']
```
Comment (nishantSharma459) [inline managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js:530]: Is this necessary?

File: managed/ui/src/components/universes/UniverseForm/ClusterFields.js (hunk 2 above)
Comment (gaurav061) [reply managed/ui/src/components/universes/UniverseForm/ClusterFields.js]: Done

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js (hunk 1 above)
Comment (gaurav061) [reply managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js:64]: This is used at multiple places, This extra param is not needed for most of them.

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js (hunk 3 above)
Comment (gaurav061) [reply managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js:530]: This is required to make async validate function call for specified fields

**Review by nishantSharma459 — APPROVED** [review]

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js [hunk 4]
```diff
@@ -447,7 +452,7 @@ const asyncValidate = (values, dispatch) => {
       values.formType !== 'Async'
     ) {
       dispatch(checkIfUniverseExists(values.primary.universeName)).then((response) => {
-        if (response.payload.status !== 200 && values.formType !== 'Edit') {
+        if (response.payload.status === 200 && values.formType !== 'Edit') {
           reject({ primary: { universeName: 'Universe name already exists' } });
```
Comment (jaydeepkumara) [inline managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js:456]: Error message on 200 status?

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js (hunk 1 above)
Comment (sshev) [inline managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js:64]:
```suggestion
    submitConfigureUniverse: (values, universeUUID = null) => {
```

File: managed/src/main/resources/v1.routes [hunk 5]
```diff
@@ -148,7 +148,7 @@ POST    /customers/:cUUID/universes/:uniUUID/upgrade                           c
 POST    /customers/:cUUID/universes/:uniUUID/disk_update                       com.yugabyte.yw.controllers.UniverseController.updateDiskSize(cUUID: java.util.UUID, uniUUID: java.util.UUID)
 GET     /customers/:cUUID/universes/:uniUUID/status                            com.yugabyte.yw.controllers.UniverseController.status(cUUID: java.util.UUID, uniUUID: java.util.UUID)
 GET     /customers/:cUUID/universes/:uniUUID/health_check                      com.yugabyte.yw.controllers.UniverseController.healthCheck(cUUID: java.util.UUID, uniUUID: java.util.UUID)
-GET     /customers/:cUUID/universes/find/:universeName                         com.yugabyte.yw.controllers.UniverseController.findByName(cUUID: java.util.UUID, universeName: String)
+GET     /customers/:cUUID/universes/find/:universeName                         com.yugabyte.yw.controllers.UniverseController.findByName_Corrected(cUUID: java.util.UUID, universeName: String)
```
Comment (sshev) [inline managed/src/main/resources/v1.routes:151]: is this change related to the AZ rendering issue?

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js [hunk 6]
```diff
@@ -61,9 +61,14 @@ import { toast } from 'react-toastify';
 
 const mapDispatchToProps = (dispatch) => {
   return {
-    submitConfigureUniverse: (values) => {
+    submitConfigureUniverse: (values, uuid = null) => {
       dispatch(configureUniverseTemplateLoading());
       return dispatch(configureUniverseTemplate(values)).then((response) => {
+        if(response.error && uuid) {
```
Comment (sshev) [inline managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js:67]:
```suggestion
        if (response.error && universeUUID) {
```

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js [hunk 7]
```diff
@@ -61,9 +61,14 @@ import { toast } from 'react-toastify';
...
+          dispatch(fetchUniverseInfo(uuid)).then((response) => {
+            dispatch(fetchUniverseInfoResponse(response.payload));
```
Comment (sshev) [inline managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js:69]: `submitConfigureUniverse()` expected to return a promise at `createUniverse()`
pls make sure it works as expected after this change

File: managed/src/main/resources/v1.routes (hunk 5 above)
Comment (gaurav061) [reply managed/src/main/resources/v1.routes:151]: No, This is an additional request to change the API of the universe name, for now, it is checking the name and if it got a response other than 200 then it says that Universe already exists.

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js (hunk 7 above)
Comment (gaurav061) [reply managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js:69]: Yes, I was able to create a universe with this change in place, But will do it once more.

**Review by jaydeepkumara — APPROVED** [review]

File: managed/src/test/java/com/yugabyte/yw/controllers/UniverseControllerTest.java [hunk 8]
```diff
@@ -1425,6 +1424,26 @@ public void testResetVersionUniverse() {
   public void testFindByNameWithUniverseDoesNotExist() {
     createUniverse(customer.getCustomerId());
     String url = "/api/customers/" + customer.uuid + "/universes/find/FakeUniverse";
+    Result result = assertThrows(YWServiceException.class,
+    () -> doRequestWithAuthToken("GET", url, authToken)).getResult();
```
Comment (Arnav15) [inline managed/src/test/java/com/yugabyte/yw/controllers/UniverseControllerTest.java:1428]: Please format.

File: managed/src/main/resources/v1.routes [hunk 9]
```diff
@@ -148,7 +148,8 @@ POST    /customers/:cUUID/universes/:uniUUID/upgrade                           c
...
+GET     /customers/:cUUID/universes/verify/:universeName                       com.yugabyte.yw.controllers.UniverseController.verifyUniqueName(cUUID: java.util.UUID, universeName: String)
```
Comment (Arnav15) [inline managed/src/main/resources/v1.routes:152]: Change name to checkUnique

**Review by Arnav15 — CHANGES_REQUESTED** [review]

File: managed/src/main/java/com/yugabyte/yw/common/kms/util/EncryptionAtRestUtil.java [hunk 10]
```diff
@@ -233,13 +233,13 @@ public static void activateKeyRef(UUID universeUUID, UUID configUUID, byte[] key
         Base64.getEncoder().encodeToString(keyRef));
   }
 
-    public static KmsHistory getActiveKeyOrBadRequest(UUID universeUUID) {
-      KmsHistory activeKey = getActiveKey(universeUUID);
-      if(activeKey == null) {
-        throw new YWServiceException(BAD_REQUEST, "Could not retrieve ActiveKey");
-      }
-      return activeKey;
+  public static KmsHistory getActiveKeyOrBadRequest(UUID universeUUID) {
```
Comment (jaydeepkumara) [inline managed/src/main/java/com/yugabyte/yw/common/kms/util/EncryptionAtRestUtil.java:236]: No code changes. Auto formate using javafmtAll.

File: managed/src/main/java/com/yugabyte/yw/controllers/EncryptionAtRestController.java [hunk 11]
```diff
@@ -36,65 +36,59 @@
 import java.util.stream.Collectors;
 
 public class EncryptionAtRestController extends AuthenticatedController {
-    public static final Logger LOG = LoggerFactory.getLogger(EncryptionAtRestController.class);
```
Comment (jaydeepkumara) [inline managed/src/main/java/com/yugabyte/yw/controllers/EncryptionAtRestController.java:39]: No code changes. Auto formate using javafmtAll.

File: managed/src/test/java/com/yugabyte/yw/controllers/EncryptionAtRestControllerTest.java [hunk 12]
```diff
@@ -154,53 +154,60 @@ public void testDeleteConfig() {
     assertAuditEntry(1, customer.uuid);
   }
 
-    @Ignore("This test passes locally but fails on Jenkins due to Guice not injecting mocked ApiHelper for an unknown reason")
```
Comment (jaydeepkumara) [inline managed/src/test/java/com/yugabyte/yw/controllers/EncryptionAtRestControllerTest.java:157]: No code changes. Auto formate using javafmtAll.

**Review by Arnav15 — APPROVED** [review]

File: managed/src/main/resources/v1.routes (hunk 5 above)
Comment (sb-yb) [reply managed/src/main/resources/v1.routes:151]: Actually leave out this change. If UI is not going to use this then i can do with other related changes.

**Review by sb-yb — CHANGES_REQUESTED** [review]
Are there any usages of  findByName or findByName_Corrected left in UI after this change?

File: managed/src/main/resources/v1.routes [hunk 13]
```diff
@@ -148,7 +148,8 @@ POST    /customers/:cUUID/universes/:uniUUID/upgrade                           c
 POST    /customers/:cUUID/universes/:uniUUID/disk_update                       com.yugabyte.yw.controllers.UniverseController.updateDiskSize(cUUID: java.util.UUID, uniUUID: java.util.UUID)
 GET     /customers/:cUUID/universes/:uniUUID/status                            com.yugabyte.yw.controllers.UniverseController.status(cUUID: java.util.UUID, uniUUID: java.util.UUID)
 GET     /customers/:cUUID/universes/:uniUUID/health_check                      com.yugabyte.yw.controllers.UniverseController.healthCheck(cUUID: java.util.UUID, uniUUID: java.util.UUID)
-GET     /customers/:cUUID/universes/find/:universeName                         com.yugabyte.yw.controllers.UniverseController.findByName(cUUID: java.util.UUID, universeName: String)
+GET     /customers/:cUUID/universes/find/:universeName                         com.yugabyte.yw.controllers.UniverseController.findByName_Corrected(cUUID: java.util.UUID, universeName: String)
+GET     /customers/:cUUID/universes/checkUnique/:universeName                  com.yugabyte.yw.controllers.UniverseController.verifyUniqueName(cUUID: java.util.UUID, universeName: String)
```
Comment (mjoshi0923) [inline managed/src/main/resources/v1.routes:152]: Do we use camelCase in rest endpoints? talking about **checkUnique** not universeName, variables are fine.

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java [hunk 14]
```diff
@@ -128,6 +127,22 @@ public Result findByName_Corrected(UUID customerUUID, String universeName) {
...
+   * @return 200 if universe is not present else raise error
+   */
+  public Result verifyUniqueName(UUID customerUUID, String universeName) {
+    // Verify the customer with this universe is present.
+    Customer.getOrBadRequest(customerUUID);
+    LOG.info("Finding Universe with name {}.", universeName);
+    if (Universe.checkIfUniverseExists(universeName)) {
+      throw new YWServiceException(BAD_REQUEST, "Universe already exists");
+    } else {
+      return withMessage("Universe does not Exist");
```
Comment (sb-yb) [inline managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:142]: Just from code readability POV. Message should be positive sounding since you are returning success. 
"Universe name is unique"

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java [hunk 15]
```diff
@@ -128,6 +127,22 @@ public Result findByName_Corrected(UUID customerUUID, String universeName) {
     return Results.status(OK, Json.toJson(universe.universeUUID));
   }
 
+  /**
```
Comment (sb-yb) [inline managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:130]: Update PR description to explain why findByName_corrected cannot be used by UI.
Looking up resource (find) is a commonly exposed API and not verifyUnique.

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java [hunk 16]
```diff
@@ -99,7 +99,6 @@ public UniverseController(YBClientService service) {
    * @return true if universe already exists, false otherwise
    */
   // TODO(.*endra):  This method is buggy. The javadoc does not match impl.
-  //  Replace findByName with findByName_Corrected when UI is fixed to expect opposite.
   @Deprecated
   public Result findByName(UUID customerUUID, String universeName) {
```
Comment (sb-yb) [inline managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:103]: - Delete this method.
- Rename findByName_Corrected to findByName
- Fix all the tests.

File: managed/src/main/resources/v1.routes (hunk 5 above)
Comment (sb-yb) [reply managed/src/main/resources/v1.routes:151]: Leave this as is. Instead rename findByName_Corrected as suggested above.

**Review by sb-yb — CHANGES_REQUESTED** [review]

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js (hunk 4 above)
Comment (sb-yb) [reply managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js]: What was wrong with this approach? It makes sense if you change the function name:
```
dispatch(findUniverseByName(values.primary.universeName)).then((response) => {
     if (response.payload.status === 200 && values.formType !== 'Edit') {
          reject()
```

**Comment by sb-yb** [comment]
Also add unit tests for new API. I don't see any.

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js (hunk 4 above)
Comment (sb-yb) [reply managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js]: Consider this. Before making any change that I suggested in comments. @gaurav061 @Arnav15

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java (hunk 14 above)
Comment (jaydeepkumara) [reply managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java]: good point.

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java (hunk 15 above)
Comment (jaydeepkumara) [reply managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:112]: Added

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java (hunk 16 above)
Comment (jaydeepkumara) [reply managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java]: 1, 2 updated. 3rd already covered. Have updated the existing test case for find API and added 2 test case for check_unique as well.

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js (hunk 4 above)
Comment (sb-yb) [reply managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js]: Also @jaydeepkumara I was hoping you would consider this before making the changes. I did not tag you because I thought that Gaurav was making these changes.

Comment (sb-yb) [reply managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js]: Will following address your concern? (See change to NOT 404 instead of 200.)
```
dispatch(findUniverseByName(values.primary.universeName)).then((response) => {
     if (response.payload.status !== 404 && values.formType !== 'Edit') {
          reject()
```
@jaydeepkumara @Arnav15

File: managed/src/main/resources/v1.routes [hunk 17]
```diff
@@ -134,6 +134,7 @@ POST    /customers/:cUUID/universe_resources                                   c
 PUT     /customers/:cUUID/universes/:uniUUID                                   com.yugabyte.yw.controllers.UniverseController.update(cUUID: java.util.UUID, uniUUID: java.util.UUID)
 POST    /customers/:cUUID/universes/import                                     com.yugabyte.yw.controllers.ImportController.importUniverse(cUUID: java.util.UUID)
 GET     /customers/:cUUID/universes                                            com.yugabyte.yw.controllers.UniverseController.list(cUUID: java.util.UUID)
+GET     /customers/:cUUID/universes/find                                       com.yugabyte.yw.controllers.UniverseController.find(cUUID: java.util.UUID)
```
Comment (jaydeepkumara) [inline managed/src/main/resources/v1.routes:137]: Move up because API URL was clashing with /customers/:cUUID/universes/:uniUUID.

**Comment by sb-yb** [comment]
Thanks for making the change. I have just one comment. Use `Collections.singletonList`  to be consistent.

**Review by sb-yb — APPROVED** [review]

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js (hunk 1 above)
Comment (gaurav061) [reply managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js]: Done

File: managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js (hunk 6 above)
Comment (gaurav061) [reply managed/ui/src/components/universes/UniverseForm/UniverseFormContainer.js]: Done.

**Review by sshev — APPROVED** [review]
UI part looks good to me

**Comment by sb-yb** [comment]
Whats holding back this PR? @gaurav061 @jaydeepkumara

**Comment by gaurav061** [comment]
@sb-yb We have a test scenario when we update the instance type of the universe, the process is getting stuck at 53% and eventually failing. @RP1807 is helping us with this test once done we will merge this.
